### DISCUSSION TITLE
Pagination offsets

### DISCUSSION THREAD

**Discussion by ramimoghrabi** [sbertix/Swiftagram#239] [Q&A] [answered]
Hello bro,

I was upgrading my project with the latest changes you have done on the api, i can see that you changed how pagination works, now i use pages(count, offset) lets say i specified one page as count, and i want to add another page when i reach the last post only, how can i accomplish this? i tried incrementing pages , but posts will duplicate in this case. Can you help me out on this ? Thanks in advance

**Comment by sbertix** [comment, marked as answer]
Hey @ramimoghrabi, 

Let's assume, for simplicity, you're fetching the first page of followers like this. 

```swift
let secret: Secret = /* a valid `Secret` */
var bin: Set<AnyCancellable> = /* a **retained** `Set` */

Endpoint.user(secret.identifier)
   .followers
   .unlock(with: secret)
   .session(.instagram)
   .pages(1)
   .catch { _ in Empty() }
   .sink {
      // Do whatever you need to do with your response here…

      guard let nextOffset = $0.offset else { return }
      // If next offset is non-`nil`, you can just cache it or
      // persist it somewhere.
   }
   .store(in: &bin)
```

Now assuming you saved `nextOffset` somewhere, when you need to load the second page of followers, you can just do:

```swift
let nextOffset: String = /* a valid `String`, obtained like above */
let secret: Secret = /* a valid `Secret` */
var bin: Set<AnyCancellable> = /* a **retained** `Set` */

Endpoint.user(secret.identifier)
   .followers
   .unlock(with: secret)
   .session(.instagram)
   .pages(1, offset: nextOffset)
   .catch { _ in Empty() }
   .sink { _ in }
   .store(in: &bin)
```

**Reply by ramimoghrabi** [reply]
Thank you for detailed explanation and example, i appreciate that.

**Reply by diligiant** [reply]
@sbertix using your guidelines for
```
Endpoint.saved.collection(id)
        .unlock(with: secret)
        .session(.instagram)
        .pages(1, offset: offset)
```
1. The first time, I provide a nil optional and get the 20 first posts
2. The second time, I provide "$0.offset" and get the 20 first posts
3. etc

I checked $0 for the presence of offset (nextMaxId), my call parameter, etc…

Any rookie mistake you'd have in mind? thank you.

**Reply by sbertix** [reply]
Hey @diligiant, assuming you're getting the "offeset" to use from the first call, this might be related to the changes Instagram has made to the colelction/autocollection response that you pointed out in another discussion/issue… 🤔 
I'm not sure when I'll get to take a look at this though, unfortunately

**Reply by diligiant** [reply]
Hey @sbertix that was quick! If tips are possible in a few lines (for your time), I'll be happy to dig in for you and report back.

**Reply by diligiant** [reply]
@sbertix I compared with Proxyman the API calls between Swiftagram & Grids (old v6 [5/2020])

The difference being that Swiftagram escapes the two last characters of max_id (always == as it seems to be a base64 string) as (%3D%3D).

**Reply by sbertix** [reply]
> @sbertix I compared with Proxyman the API calls between Swiftagram & Grids (old v6 [5/2020])
> 
> 
> 
> The difference being that Swiftagram escapes the two last characters of max_id (always == as it seems to be a base64 string) as (%3D%3D).

We're talking of the request body, not the query items, correct? 🤔

**Reply by diligiant** [reply]
(didn't notice at first : there is no `all` after the collectionId in grids)
grids:
`https://i.instagram.com/api/v1/feed/collection/17853802247011558/?max_id=QVFCSTV1NkVwVFpINTJjUUNpVG9RUzdVOHVTSmRYek1tZzBfaDVXbjc2OXJnQUR1bnNURmxBQnlfOEFSeDJmR2gtRWhZUThxWTR3SFRvSWctUnNXdl96Mw==`

swiftagram:
`https://i.instagram.com/api/v1/feed/collection/17907088288931602/all/?show_igtv_first=false&max_id=QVFCeHdMS1hEaF9oLUhRRjdXbkJwb2tpNHozZGVZVnFGd3NoVE1qU09vRkNlSENHWFdhdjZDUThMMFRlUkpsNFpoR2ZDUnJsRHpXVTFWRklJSk1hSHFSRg%3D%3D&include_igtv_preview=true`

I am installing a proxy on an Android to check what the native app is doing.

**Reply by sbertix** [reply]
> (didn't notice at first : there is no `all` after the collectionId in grids)
> 
> grids:
> 
> `https://i.instagram.com/api/v1/feed/collection/17853802247011558/?max_id=QVFCSTV1NkVwVFpINTJjUUNpVG9RUzdVOHVTSmRYek1tZzBfaDVXbjc2OXJnQUR1bnNURmxBQnlfOEFSeDJmR2gtRWhZUThxWTR3SFRvSWctUnNXdl96Mw==`
> 
> 
> 
> swiftagram:
> 
> `https://i.instagram.com/api/v1/feed/collection/17907088288931602/all/?show_igtv_first=false&max_id=QVFCeHdMS1hEaF9oLUhRRjdXbkJwb2tpNHozZGVZVnFGd3NoVE1qU09vRkNlSENHWFdhdjZDUThMMFRlUkpsNFpoR2ZDUnJsRHpXVTFWRklJSk1hSHFSRg%3D%3D&include_igtv_preview=true`
> 
> 
> 
> I am installing a proxy on an Android to check what the native app is doing.

I guess it makes more sense for the `all` to be what's causing the issue (assuming it's not just the wrong offset being returned by Swiftagram's model). I doubt it's about escaping query items (I don't really know what Grids is but they would still need to escape it to have the request "work", it's probably just applied after the "logging phase", and for sure Insta unescapes it correctly)…

**Reply by diligiant** [reply]
You're right, that's not the issue (Grids is a MacOS insta client).

I changed `.path(appending: "all/")` to `.path(appending: "/")` in `Endpoint.Group.Saved.Collection.summary` and was able to "page". The returned structure is a little different that current Unit (the elements of `self["saveMediaResponse"]` are top level.)

When I compare all/ and /, the former returns a lot of information helping the official client to layout its user interface. As such, it seems to me that it should be kept ( all/ => without paging) and another api (/ => with paging) should appear. It that the spirit for you?

**Reply by sbertix** [reply]
> You're right, that's not the issue (Grids is a MacOS insta client).
> 
> 
> 
> I changed `.path(appending: "all/")` to `.path(appending: "/")` in `Endpoint.Group.Saved.Collection.summary` and was able to "page". The returned structure is a little different that current Unit (the elements of `self["saveMediaResponse"]` are top level.)
 
Great 💪 thx man.
 
> When I compare all/ and /, the former returns a lot of information helping the official client to layout its user interface. As such, it seems to me that it should be kept ( all/ => without paging) and another api (/ => with paging) should appear. It that the spirit for you?

I'm personally not a big fan of duplication.  
I'd rather fix `/all` to have it paginate, especially if `/` is just legacy and never actually called in the client (mainly for spam detection). Were you checking that too? Is the app actually ever requesting the non-`/all` endpoint?

**Reply by diligiant** [reply]
I might have badly explained:

`all/` (your current one) returns in saveMediaResponse what `/` returns top-level. `/` is the one (new) we need to be able to page (therefore `all/` returning me the first elements indefinitely ;)

BUT `all/` is not "deprecated". It just returns more elements that are mainly of use by people willing to emulate the Instagram UI. I guess the official application first calls `all/` and then subsequently pages with `/`.

If it were me (but it's not), I'd keep the existing `all/` and create a new `/` call (The most difficult part is coming with good names…)

**Reply by sbertix** [reply]
Sorry, I think I messed up with submitting the replying cause I don't see it anymore. 
What I meant, as you mentioned inspecting the Android app's requests, was wether calling `/` and `/all` was a behaviour existing in the app or not. Cause I'm sure `/all` was called when I implemented it (as I got it from there), so I was wondering if perhaps `/` was a deprecated endpoint, never called in the client, meaning we shouldn't implement it in **Swiftagram** and that there's probably a way to make `/all` work with offsets instead. 

>  I guess the official application first calls all/ and then subsequently pages with /.

So if it's just an educated guess I'll try and actually check the requests. No promises on when though. 

> If it were me (but it's not), I'd keep the existing all/ and create a new / call (The most difficult part is coming with good names…)

I'd rather mimic the design of the timeline `Endpoint`, as I don't expect users to actually value them separately: calling `/all` with a `nil` offset and `/` in any other case, even on "automatic" pagination.

**Comment by ramimoghrabi** [comment]
one more thing i wanted to know about timeline, that 
Endpoint.posts.recent
                .unlock(with: secret!)
                .session(.instagram)
                .pages(1)
.sink(receiveCompletion: {print ("completion: \($0)")}, receiveValue: { (wrap) in
                    guard let nextOffset1 = wrap.offset else { return }  ---> Initializer for conditional binding must have Optional type, not 'Wrapper'                   
})


does not work as described above ? does it have another approach ?

**Reply by sbertix** [reply]
The timeline endpoint works a bit differently cause it encompasses many exceptions. 

You can check offset computation directly from its definition in code, but I'll just add it here for you.

```swift
let output: Wrapper = /* a valid `Wrapper` you get as `Endpoint.recent.posts` output */

let nextOffset: String?
switch output.nextMaxId.string(converting: true) {
case .none:
    nextOffset = nil
case "feed_recs_head_load":
    nextOffset = output.feedItems
                .array()?
                .last?
                .endOfFeedDemarcator
                .groupSet
                .groups
                .array()?
                .first(where: { $0.id.string(converting: true) == "past_posts" })?
                .nextMaxId
                .string()
case let cursor?:
    nextOffset = cursor
}
```

**Reply by ramimoghrabi** [reply]
again i wanna thank you for giving me part of your valuable time.

**Comment by ramimoghrabi** [comment]
i think there might be a bug , in comments offset, setting the offset on first fetch is always nil, thus it won't load since i used the above return method if not found : guard let nextOffset = $0.offset else { return }

**Reply by sbertix** [reply]
I'm assuming the post you mentioned had more than one page of comments, correct? 
(Consider threaded comments do not appear in pagination as they have their own offset)

**Reply by ramimoghrabi** [reply]
yeah i want to load more than page and the post has many pages of comments, can you guide just to the offsets of comments ? just the title where to read about it or find it if you don't have time for examples

**Reply by sbertix** [reply]
Unfortunately I can't check it tonight. I'm definitely gonna take a look at it and fix it, if it is a bug, tomorrow 💪

**Reply by ramimoghrabi** [reply]
sure no problem, thanks

**Reply by ramimoghrabi** [reply]
Endpoint.users(matching: query)
            .unlock(with: secret)
            .session(.instagram)
            .pages(1)
            .sink(receiveCompletion: {_ in }, receiveValue: { users in
                guard let nextOffset1 = users.offset else { return }
})

same issue here offset is nil. i am just testing and mentioning them

**Reply by sbertix** [reply]
Sorry I didn't manage to come back to this: busy week. 
I'm planning on checking this asap and definitely fix it, if needs to, by the end of the week at the latest. 
Apologies @ramimoghrabi

**Reply by ramimoghrabi** [reply]
no problem brother, just let me know when it is fixed so that i can test it again.

**Comment by ramimoghrabi** [comment]
hello again @sbertix ,i was wondering if you had time to check the comments pagination, as the wrapper.offset is always nil.
